traffic_ops/app/db/reencrypt/reencrypt.go

```diff
@@ -0,0 +1,393 @@
+
+Description
+  The reencrypt app is used to re-encrypt all data in the Postgres Traffic Vault
+  using a new base64-encoded AES key.
+
+Options
+	--previous-key
+        The file path for the previous base64-encoded AES key.
+
+	--new-key
+        The file path for the new base64-encoded AES key.
+
+	--cfg
+        The path for the configuration file. Default is `./reencrypt.conf`.
+
+*/
+
+package main
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import (
+	"crypto/aes"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/apache/trafficcontrol/lib/go-util"
+
+	"github.com/jmoiron/sqlx"
+	_ "github.com/lib/pq"
+)
+
+const PROPERTIES_FILE = "./reencrypt.conf"
+
+func main() {
+	previousKeyLocation := flag.String("previous-key", "", "The file path for the previous base64 encoded AES key.")
+	newKeyLocation := flag.String("new-key", "", "The file path for the new base64 encoded AES key.")
+	cfg := flag.String("cfg", PROPERTIES_FILE, "The path for the configuration file. Default is "+PROPERTIES_FILE+".")
+	flag.Parse()
+
+	if len(os.Args) < 4 {
+		flag.Usage()
+		os.Exit(1)
+	}
+
+	if previousKeyLocation == nil || *previousKeyLocation == "" {
+		die("previous-key flag is required.")
+	}
+	if newKeyLocation == nil || *newKeyLocation == "" {
+		die("new-key flag is required.")
+	}
+
+	newKey, err := readKey(*newKeyLocation)
+	if err != nil {
+		die("reading new-key: " + err.Error())
+	}
+
+	previousKey, err := readKey(*previousKeyLocation)
+	if err != nil {
+		die("reading previous-key: " + err.Error())
+	}
+
+	dbConfBytes, err := ioutil.ReadFile(*cfg)
+	if err != nil {
+		die("reading db conf '" + *cfg + "': " + err.Error())
+	}
+
+	pgCfg := Config{}
+	err = json.Unmarshal(dbConfBytes, &pgCfg)
+	if err != nil {
+		die("unmarshalling '" + *cfg + "': " + err.Error())
+	}
+
+	sslStr := "require"
+	if !pgCfg.SSL {
+		sslStr = "disable"
+	}
+	db, err := sqlx.Open("postgres", fmt.Sprintf("postgres://%s:%s@%s:%d/%s?sslmode=%s&fallback_application_name=trafficvault", pgCfg.User, pgCfg.Password, pgCfg.Hostname, pgCfg.Port, pgCfg.DBName, sslStr))
+	if err != nil {
+		die("opening database: " + err.Error())
+	}
+
```

Comment:
Rather than creating a transaction per table, we should create a single transaction here and use it for everything. That way, if any errors occur, the _entire_ change is rolled back, and we should never end up in a state where the TV DB was partially re-encrypted. That would be a bit of a mess 😃 .

Also, if an error occurs, we should rollback the transaction instead of committing, and if the transaction commit/rollback is deferred in this function, any paths that lead to calling `die` will skip any deferred functions.